### PR TITLE
[java] Revamp GuardLogStatementRule to allow var, field and array accesses

### LINKED ISSUE #3602: [java] GuardLogStatement: False positive when compile-time constant is created from external constants
**Affects PMD Version:**
6.40.0+
7.0.0

**Rule:** [GuardLogStatement](https://pmd.github.io/pmd-6.40.0/pmd_rules_java_bestpractices.html#guardlogstatement)

**Description:**
The changes to resolve issue #957 work well when the constant is located in the same class (or at least the same file -- I did not try the scenario of different classes and same file).  However, when the concatenated constant used to create the compile-time constant log message comes from elsewhere, PMD still issues a GuardLogStatement false positive.

**Code Sample demonstrating the issue:**
```java
public class OtherClass {
  public static final String TERM_MSG = "A terminating log message.";
}
```
```java
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class MyClass {
  private static final Logger logger = LogManager.getLogger();
  public void doSomething() {
     final Object arg = new Object();
     logger.info("Problem message: {}. " + OtherClass.TERM_MSG, arg);
  }
}
```

**Expected outcome:**
Because the two concatenated constants result in a compile time constant, there is no work done at log time to create the message string if the log level does not include INFO and so there is no need for a guard statement.  This is the same scenario as #957 except in this case, the concatenated constant String is not coming from within the same class.

PMD reports a GuardLogStatement violation at the `logger.info(...)` line, but that's wrong. That's a false positive.

**Running PMD through:** Eclipse PMD Plugin 4.29

### LINKED ISSUE #4731: [java] GuardLogStatement: Documentation is unclear why getters are flagged
**Affects PMD Version:** 
7.0.0-rc4

**Rule:**
[GuardLogStatement](https://docs.pmd-code.org/pmd-doc-7.0.0-rc4/pmd_rules_java_bestpractices.html#guardlogstatement): Logger calls should be surrounded by log level guards.

**Description:**
```java
            if (handler.isPresent()) {
                handlerMap.put(field, handler.get());
            } else {
                logger.error("Unsupported field: '{}.{}'", type.getId(), field.getName());
            }
        }

```

other example:
```java
            if (result == null) {
                logger.warn("Unknown enum value: '{}' for enum {}", data.getAsString(), enumType.getSimpleName());
            }

```

**Expected outcome:**

PMD reports a violation at line "logger.error", the example in the documentation says: no guard necessary if parameters are there; getters are parameters.

**Running PMD through:** *Gradle*

### LINKED ISSUE #5151: [java] GuardLogStatement: Should not need to guard parameterized log messages where the replacement arg is a constant from another class
**Affects PMD Version:** 
7.0.0

**Rule:**
[GuardLogStatement](https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement)

**Description:**
I do not believe an argument, which is a constant from another class, should result in a violation of this rule.

**Code Sample demonstrating the issue:**

```java
        // This IS flagged as a violation
        LOG.info(
                "Some message here : foo={}",
                SomeOtherClass.MY_CONSTANT
        );

        // This is not flagged as a violation
        LOG.info(
                "Some message here : foo={}",
                MY_CONSTANT
        );
```

**Expected outcome:**

PMD reports "... Rule:GuardLogStatement Priority:2 Logger calls should be surrounded by log level guards..".
I think that's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Run via maven-pmd-plugin (3.24.0)

### LINKED ISSUE #5152: [java] GuardLogStatement: Should not need to guard parameterized log messages where the replacement arg is "this"
**Affects PMD Version:** 
7.0.0

**Rule:**
[GuardLogStatement](https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement)

**Description:**
I do not believe an argument, which is `this`, should result in a violation of this rule.

**Code Sample demonstrating the issue:**

```java
        LOG.info(
                "Some message here : foo={}",
                this
        );
```

**Expected outcome:**

PMD reports "... Rule:GuardLogStatement Priority:2 Logger calls should be surrounded by log level guards..".
I think that's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Run via maven-pmd-plugin (3.24.0)

### LINKED ISSUE #5153: [java] GuardLogStatement: Should not need to guard parameterized log messages where the replacement arg is an array element
**Affects PMD Version:** 
7.0.0

**Rule:**
[GuardLogStatement](https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement)

**Description:**
I do not believe an argument, which is an array element, should result in a violation of this rule.

**Code Sample demonstrating the issue:**

```java
        final String[] foo = ...
        LOG.info(
                "Some message here : foo0={}",
                foo[0]
        );
```

**Expected outcome:**

PMD reports "... Rule:GuardLogStatement Priority:2 Logger calls should be surrounded by log level guards..".
I think that's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Run via maven-pmd-plugin (3.24.0)

### PR DESCRIPTION
**Pull request by jsotuyod**
## Related issues

- Fixes #3602 
- Fixes #4731
- Fixes #5151
- Fixes #5152
- Fixes #5153

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 13 violations, 2 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/10444229280?pr=5169">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java [hunk 1]
```diff
@@ -237,7 +240,14 @@ private boolean isUnguardedAccessOk(ASTMethodCall call, int messageArgIndex) {
...
                    .drop(messageArgIndex) // remove the level argument if needed
-                   .all(it -> it instanceof ASTStringLiteral || it instanceof ASTLambdaExpression || it instanceof ASTVariableAccess || it instanceof ASTMethodReference);
+                   .all(GuardLogStatementRule::isDirectAccess);
+    }
+
+    private static boolean isDirectAccess(ASTExpression it) {
+        return it instanceof ASTStringLiteral || it instanceof ASTLambdaExpression
+                || it instanceof ASTVariableAccess || it instanceof ASTMethodReference
+                || it instanceof ASTFieldAccess || it instanceof ASTThisExpression
+                || (it instanceof ASTArrayAccess && isDirectAccess(((ASTArrayAccess) it).getQualifier()));
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java:250]: Checking that the qualifier is a direct access should also be done for method references and field accesses (all QualifiableExpressions) : `foo().field`, `foo()::ref`

**Comment by jbisotti** [comment]
It would be great if the documentation could also be expanded to explain the nuances here. Maybe include examples like in my original Issues that should pass the check and then slight variations of them that would not. Thanks.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java:250]: I think it's illegal to use anything else than a class type as the LHS of a this expression, so I'm not sure it's worth checking for that

**Comment by jsotuyod** [comment]
@jbisotti I changed the rule description slightly and added some extra examples, feel free to comment.

The bottom line is, we should flag any "expensive operations", with expensive being one of:
- string manipulation / concatenation
- method invocations
- constructor invocations

The last 2 are specially dangerous as the methods may also have side-effects, so you are using a log call to actually perform business logic.

Accessing fields / variables / array elements / method references directly is ok. Doing that from the return of a method invocation / constructor call is not.

**Review by adangel — APPROVED** [review]
Thanks!

**Pull request merged**